Fix #5701: Alert user to save changes before publishing

ankita240796 opened the pull request:
## Explanation
Fixes #5701 
Added mechanism to alert user if user clicks on publish button without saving draft.

Here is a screenshot for the same:
![image](https://user-images.githubusercontent.com/15226041/49570864-c22ffe00-f95d-11e8-9ab3-6782ad02d679.png)

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.

ankita240796 committed:
Alert user to save changes before publishing

ankita240796 commented:
Hi @tjiang11, PTAL!

DubeySandeep commented:
Hi @ankita240796, Thanks for the PR! 

@nithusha21, can you please review this PR as I'm not sure about the expected behaviour for the issue #5701.

nithusha21 commented:
Hi @ankita240796 Thanks for the PR. 

I am a little unsure myself about why this issue was filed. I think the issue was probably clicking a greyed out publish button, which was actually not clickable. In which case I think your fix is good.

However, I would suggest changing the error message to also include some context regarding open warnings that are not yet resolved. So if the creator saves the draft, and still there are open warnings, the publish button will still be disabled, and the pop-up still asks me to save draft (which I cannot anymore).

Probably, a better UX would be to have separate messages for either situation. So the current message is shown only when there are no open warnings and the draft isn't saved. Else, if there are warnings, we should show a message saying that. What do you think of this?

ankita240796 commented:
Hi @nithusha21, I totally agree with your suggestion. It would be better to show the relevant error message to the user as per the situation. I will make the changes for the same.
Thanks!

ankita240796 committed:
Update warning message

ankita240796 commented:
Hi @nithusha21, I have made the changes, PTAL!

seanlip commented:
Hi all, I don't have enough context to comment usefully on the whole PR, but one thing I wanted to mention is that we never use in-built alert boxes -- they don't look great. Modals are the way to go here, as per other similar user interactions in Oppia.

Thanks!

nithusha21 commented:
Hi Sean, thanks for those inputs! Yeah, I would agree with Sean. @ankita240796 could you make these alerts into Modals? You could search for uibModal in the codebase to get examples of how to create a modal. Let me know if you need help!

bansalnitish commented:
Yeah, I too agree with Sean. We have never used in-built alerts. Let us know if you need any help. More than happy to help :)

ankita240796 commented:
Hi @nithusha21 @seanlip @bansalnitish , Would it be better if instead of creating a modal we just try to display the tooltip when the button is clicked?

ankita240796 committed:
Added info modal for publish button

ankita240796 commented:
Hi @nithusha21 @seanlip @bansalnitish 

I have created a info modal to show the warning message to user if publish button is clicked without saving the drafts or resolving the warnings.

Here is a screenshot for the case when draft is not saved:
![image](https://user-images.githubusercontent.com/15226041/49640690-389f2f80-fa34-11e8-9005-68f90a129b46.png)

Do let me know if this looks fine or should we try displaying the tooltip on click.

Also, would it be better if we provide an option to save draft in the info modal itself?

Thanks!

nithusha21 commented:
Hi @ankita240796 I think the modal looks great! Also I like the idea you have to add a button to save changes. 

One suggestion I have is to drop the ? and change the title to "Cannot publish changes.".

Also, If we are doing this, do we want to continue to have the publish button greyed out in this situation? I am not entirely sure. Any suggestions @bansalnitish, @DubeySandeep, @aks681?

DubeySandeep commented:
@nithusha21:  

I think a grey button explains the button is disabled If it's disabled we should not allow them to be clickable (we can use [this](https://proxy.duckduckgo.com/iu/?u=https%3A%2F%2Fcdn1.iconfinder.com%2Fdata%2Ficons%2Fcursor-pointers%2F24%2F10-512.png&f=1) cursor instead of hand-cursor), in that case, we should use a tooltip.

If we are going to use modal then we should enable th Publish button.

Again, we use grey color "save draft" button when there is nothing to save, What are we going to do here for a consistent UI behaviour in both buttons?

Re "add a button to save changes.":
I liked the Idea, but what we are going to do if somone clicked the "Save draft" button in the modal, will we publish the exploration after saving draft? (In that case, I think sometimes we cannot publish exploration when there are warnings in the exploration (like doesn't have End exploration card).

ankita240796 committed:
Add option to save draft in info modal

ankita240796 commented:
Hi @nithusha21 @DubeySandeep I tried out some of the stuff discussed in the comments above and have also pushed the code for the same.

Here is a gif for the same:
![save](https://user-images.githubusercontent.com/15226041/49646977-ed424c80-fa46-11e8-9c82-4c0112d0a66b.gif)

PTAL, I will change the code according to your inputs on this.

I also think that the publish button should not be greyed out if we are using a modal.

bansalnitish commented:
I agree with @DubeySandeep that the save draft and publish should be consistent. 

Also I think modals here look a bit unecessary, tooltip is a good way to display a simple message (the reason why its not working) like during publish we have the tooltip `please resolve the warnings` etc. We can add tooltip to both the save draft button and the publish button and display the tooltip accordingly like when to display `resolve warnings`, `save draft`, etc.

Any suggestions @DubeySandeep, @nithusha21, @ankita240796 ?

ankita240796 commented:
I think I agree with @bansalnitish that a modal here would be too complex. It would be better to use a tooltip. 
Also, if we are using modal for a Publish button, then we need to provide the same consistent behaviour for Save Draft button

nithusha21 commented:
Hi @ankita240796, Sorry about the confusion on this PR! This is one of those issues for which we have no "correct" solution. There are many possibilities as outlined above. Could you write these down in a doc, and compare the pros and cons of these various choices? That way we will know what would be the best alternative to take. 

Otherwise, I think every time you try something new, one of us will have a different solution, and we will not be able to finally resolve the issue! Does this sound good?

ankita240796 commented:
Sure @nithusha21, I will do that!

ankita240796 commented:
Hi @nithusha21 @DubeySandeep @bansalnitish 

Here is the doc for the approaches we have discussed till now: https://docs.google.com/document/d/16ZdhCsBOlFtPpaw5cP2ZTcSOfgZmERJk_eDVROE2GlU/edit?usp=sharing

Please provide your inputs here.

Thanks!

bansalnitish commented:
Nice work @ankita240796. I went through the doc and to me the tooltip still looks a better choice over the other (considering the pros are same and the corns for modal are high). Let's wait to listen from other two as well. Thanks!

bansalnitish commented:
@nithusha21 could you give your inputs here?

nithusha21 commented:
Hi @ankita240796 The doc looks great! Thanks!

I would lean towards the second solution, with a slight modification. Let a comment on the doc. PTAL!

ankita240796 committed:
Use tooltip and change cursor on disabled button

ankita240796 committed:
Fix minor nit

ankita240796 commented:
Hi @nithusha21 @DubeySandeep @bansalnitish 

I have made the changes as discussed on the doc. Here is working demo for the same:

![publish](https://user-images.githubusercontent.com/15226041/49735560-870b3300-fcad-11e8-887d-52fc17b16ff2.gif)

Also, I replaced the title element with tooltip since I was unable to find a way to display the title both on hover and click. This [link](https://stackoverflow.com/questions/25045607/how-to-show-gwt-hover-text-on-a-click) also mentions that trying to show title on click as well as hover runs a risk of showing two tooltips at the same time.

PTAL, Thanks!

nithusha21 commented on core/templates/dev/head/pages/exploration_editor/exploration_save_and_publish_buttons_directive.html:
```diff
@@ -1,7 +1,9 @@
-         title="<[getPublishExplorationButtonTooltip()]>">
+         uib-tooltip="<[getPublishExplorationButtonTooltip()]>"
+         tooltip-placement="bottom"
+         ng-class="{'div-disable': isExplorationLockedForEditing() || countWarnings()}" ng-click="showTooltip()">
```
is div-disable a standard class you want to use? If not, could you please find a more descriptive name. Maybe something like disable-publish-button or likewise.

ankita240796 replied:
I need to use this class both for publish and save draft buttons when they are disabled. Would it be better to use separate names for both?

nithusha21 replied:
I think separate names would be better.

ankita240796 replied:
Done.

nithusha21 commented on core/templates/dev/head/pages/exploration_editor/ExplorationSaveAndPublishButtonsDirective.js:
```diff
@@ -117,6 +117,13 @@ oppia.directive('explorationSaveAndPublishButtons', [
               });
           };
+
+          $scope.showTooltip = function() {
```
Could the function name also include some details about which Tooltip? showTooltip sounds a bit too generic.

ankita240796 replied:
Done

nithusha21 commented on core/templates/dev/head/pages/exploration_editor/exploration_save_and_publish_buttons_directive.html:
```diff
@@ -69,3 +74,9 @@
 </ul>
+
+<style>
+  .div-disable {
```
This style will apply too generally. We should scope it to this specific directive. 
```suggestion
  exploration-save-and-publish-buttons .div-disable {
```

ankita240796 replied:
Done

nithusha21 commented on core/templates/dev/head/pages/exploration_editor/ExplorationSaveAndPublishButtonsDirective.js:
```diff
@@ -117,6 +117,13 @@ oppia.directive('explorationSaveAndPublishButtons', [
           };
+
+          $scope.showTooltip = function() {
+            if ($scope.isExplorationLockedForEditing() ||
```
Break after the opening (

ankita240796 replied:
Done

nithusha21 commented on core/templates/dev/head/pages/exploration_editor/exploration_save_and_publish_buttons_directive.html:
```diff
@@ -1,7 +1,9 @@
-         title="<[getPublishExplorationButtonTooltip()]>">
+         uib-tooltip="<[getPublishExplorationButtonTooltip()]>"
+         tooltip-placement="bottom"
+         ng-class="{'div-disable': isExplorationLockedForEditing() || countWarnings()}" ng-click="showTooltip()">
```
Just a query. If the button is disabled, will the ng-click event get triggered if clicked? If not when is the showTooltip function triggered? And why do we need it?

ankita240796 replied:
The `ng-click` on the div element gets triggered since there is no disable on the div.

nithusha21 replied:
Ah I see. Thanks for the explanation. But on hover itself the tooltip is shown right? how does the click help?

ankita240796 replied:
In the current working, when the user clicks the tooltip is gone but in current scenario, the tooltip stays on click. I am not sure if this is needed since the button is disabled which already implies user cannot click on it.

nithusha21 replied:
I think the tooltip should show immediately on hover as it does. I think disappearing on click is ok. (I tried on google docs, and their tooltips disappear on click). Also, when disabled, anyway the clicks would not register I think. 

After removing, if on clicking the disabled button, the tooltip disappears, then lets keep this functionality. If on clicking the disabled button, the tooltip stays, let's get rid of the ng-clicks.

ankita240796 replied:
If that is the case, we can just use the title attribute and that will work fine. I will update the PR accordingly!

ankita240796 replied:
Done, PTAL!

nithusha21 reviewed (changes requested):
Hi @ankita240796 This looks much better. Thanks! The behavior LGTM. Just some stylistic comments, and queries. PTAL

nithusha21 commented:
Also, please note some travis failures. I think what happens is the tooltip for "save draft" stays open on clicking the dropdown arrow, and prevents selecting "discard draft". We need to fix that :)

ankita240796 commented:
> I think what happens is the tooltip for "save draft" stays open on clicking the dropdown arrow, and prevents selecting "discard draft". We need to fix that :)

Hi @nithusha21, would hiding the tooltip after a small delay be a good solution here?

ankita240796 commented:
I just figured out that changing the placement will fix the issue!

ankita240796 committed:
Address review comments

ankita240796 commented:
Hi @nithusha21, I have made the changes, PTAL!

nithusha21 commented on core/templates/dev/head/pages/exploration_editor/exploration_save_and_publish_buttons_directive.html:
```diff
@@ -32,9 +33,9 @@
-         ng-click="showTooltip()">
+         tooltip-placement="right"
+         ng-class="{'disable-save-button': !isExplorationSaveable()}"
+         ng-click="showTooltipForDisabledButtons(disable-save-button)">
```
Shouldn't disable-save-button be a string?

ankita240796 replied:
Fixed.

nithusha21 commented on core/templates/dev/head/pages/exploration_editor/ExplorationSaveAndPublishButtonsDirective.js:
```diff
@@ -118,10 +118,11 @@ oppia.directive('explorationSaveAndPublishButtons', [
-            if ($scope.isExplorationLockedForEditing() ||
+          $scope.showTooltipForDisabledButtons = function(buttonClass) {
+            if (
+              $scope.isExplorationLockedForEditing() ||
```
Can you indent this line and the next by 2 spaces? This will separate it from the following code.

ankita240796 replied:
It gives an indentation error: `Expected indentation of 14 spaces but found 16  indent`

nithusha21 replied:
Ah ok. I wasn't sure if it would do that. Fine to leave as-is.

nithusha21 commented on core/templates/dev/head/pages/exploration_editor/exploration_save_and_publish_buttons_directive.html:
```diff
@@ -3,7 +3,8 @@
          tooltip-placement="bottom"
-         ng-class="{'div-disable': isExplorationLockedForEditing() || countWarnings()}" ng-click="showTooltip()">
+         ng-class="{'disable-publish-button': isExplorationLockedForEditing() || countWarnings()}"
+         ng-click="showTooltipForDisabledButtons(disable-publish-button)">
```
ditto

ankita240796 replied:
Done

nithusha21 commented:
Also, I think the placement down looks better. Every other option has the tooltip downwards. So can we try this, When the dropdown is open, can we hide the tooltip?

ankita240796 committed:
Fix tooltip and dropdown conflict

ankita240796 commented:
Hi @nithusha21, I have made the changes as suggested as well as fixed the tooltip and dropdown conflict, PTAL!

ankita240796 committed:
Remove tooltip on click, use title instead

nithusha21 commented:
Hi Ankita, what was the reasoning behind removing the tooltip though? I thought the tooltip looked better. (also, all other buttons on the top have tooltips.). Lets do this:
- Lets keep the tooltip on the publish changes button (which was the cause of the original issue)
- The save draft can stay as a title (The button feels self explanatory, and I don't think the tooltip added too much info). 

Does this sound good?

ankita240796 commented:
Hi @nithusha21, I added the tooltip because it was easy to display that on click rather than a title attribute which was used earlier. If we just want the info on hover then a title attribute does the work, so I removed the tooltip and reverted to the original state of code.

Since the title also shows the same bubble as a tooltip and it is in bigger fonts, I feel that title looks better than a tooltip for these two buttons. What do you think?

Thanks!

nithusha21 commented:
I didn't actually try out how the title tag works. Let me try it, and i'll let you know.

ankita240796 commented:
Hi @nithusha21, here are some screenshots for clarification:

**Using tooltip:**

![image](https://user-images.githubusercontent.com/15226041/49802392-c3549700-fd72-11e8-9f6e-27c049cfbcc8.png)

**Using title:**

![image](https://user-images.githubusercontent.com/15226041/49802698-a2d90c80-fd73-11e8-9243-3b9d86462158.png)

nithusha21 commented:
Yeah I agree visually they look the same. However I seem to prefer the tooltip as it shows up instantly on hover. The title takes a couple of seconds before appearing. So I would go with the tooltip, as it seems to address the original issue. Instantly on hover, the user is prompted with why it is not clickable. Do you agree?

nithusha21 commented:
Also, regarding the click behaviour, I think it is not required. Only the hover behaviour of the tooltip looks to be useful to us. Once prompted that it is not clickable, it is very unlikely that someone will click it.

ankita240796 committed:
Replace publish title with tooltip

ankita240796 commented:
Hi @nithusha21, I agree with your opinion on title and tooltip.

> Lets keep the tooltip on the publish changes button (which was the cause of the original issue)
The save draft can stay as a title (The button feels self explanatory, and I don't think the tooltip added too much info).

I have made the changes as suggested by you, PTAL. Thanks!

nithusha21 reviewed (approved):
I think it looks great now! Thanks for fixing this :)

LGTM @ankita240796
